### PR TITLE
[BEAM-985] Fix shared state across retry decorated functions

### PR DESCRIPTION
**Pull request by sb2nov**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

The fix was fairly straight forward and just needed to move the generator inside the wrapped function.

R: @aaltay @chamikaramj  PTAL

### CODE REVIEW

File: sdks/python/apache_beam/utils/retry_test.py [hunk 1]
```diff
@@ -164,5 +165,50 @@ def test_log_calls_for_transient_failure(self):
       self.assertEqual(func_name, 'transient_failure')
 
 
+class DummyClass(object):
+  def __init__(self, m):
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/retry_test.py]: Could you spell out some variables
like `m` -> `mock`
`a` -> `value` ? etc.

File: sdks/python/apache_beam/utils/retry_test.py [hunk 2]
```diff
@@ -164,5 +165,50 @@ def test_log_calls_for_transient_failure(self):
...
+    self.m = m
+
+  @retry.with_exponential_backoff(num_retries=2, initial_delay_secs=0.1)
+  def func(self, a):
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/retry_test.py]: Is there a need for `a`. Why not just call `mfunc()`

File: sdks/python/apache_beam/utils/retry_test.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 """Unit tests for the retry module."""
 
 import unittest
+import mock
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/retry_test.py]: Is there a need for using `mock`

The tests could be simplified in `DummyClass` takes a list of values to return. and `func` returns or raises them one by one.

**Comment by aaltay** [comment]
Thank you @sb2nov . I have a few test comments, other than that it is good.

File: sdks/python/apache_beam/utils/retry_test.py (hunk 3 above)
Comment (sb2nov) [inline sdks/python/apache_beam/utils/retry_test.py]: Done

File: sdks/python/apache_beam/utils/retry_test.py (hunk 1 above)
Comment (sb2nov) [inline sdks/python/apache_beam/utils/retry_test.py]: Done

File: sdks/python/apache_beam/utils/retry_test.py (hunk 2 above)
Comment (sb2nov) [inline sdks/python/apache_beam/utils/retry_test.py]: Done

**Comment by sb2nov** [comment]
Thanks for the review, updated the tests to be cleaner.

R: @aaltay PTAL

**Comment by aaltay** [comment]
LGTM. Thanks.

**Comment by sb2nov** [comment]
R: @robertwb please merge.

Thank you :)

**Comment by vikkyrk** [comment]
@robertwb ping for merge. One of my datastoreio test is running into this issue too.

**Pull request closed without merging**